tests/cephadm/test_custom_container.py

```diff
@@ -0,0 +1,156 @@
+from cli.utilities.containers import Container, Registry
+from utility.log import Log
+from utility.utils import get_cephci_config
+
+log = Log(__name__)
+
+PRIVATE_REG_USER = "myregistryusername"
+PRIVATE_REG_PASS = "myregistrypassword"
+PORT = "5000"
+
+
+class ClusterCustomContainerError(Exception):
+    pass
+
+
+def get_registry_details(config):
+    """Get registry and registry credentials
+    Args:
+        config (dict): Config parameters
+    """
+    build_type = "ibm" if config.get("ibm_build") else "rh"
+    con = get_cephci_config()
+    reg_cred = con.get(f"{build_type}_registry_credentials", {})
+    return reg_cred.get("registry"), reg_cred.get("username"), reg_cred.get("password")
+
+
+def run(ceph_cluster, **kw):
+    """Verify re-deploying of monitoring stack with custom images
+    Args:
+        ceph_cluster(ceph.ceph.Ceph): CephNode or list of CephNode object
+    """
+
+    # Get installer node
+    installer = ceph_cluster.get_nodes(role="installer")[0]
+
+    # Define CephAdm class object
+    cephadm = CephAdm(installer)
+
+    # Get config file and node
+    config = kw.get("config")
+
+    # Get cluster nodes
+    ceph_nodes = kw.get("ceph_nodes")
+
+    # Get dashboard image
+    registry_image = config.get("dashboard_images")
+    dashboard_images = registry_image.values()
+
+    # Get registry docker image
+    docker_reg_image = config.get("registry_image").get("docker_registry_image")
+
+    # Get the rhbuild
+    rhbuild = config.get("rhbuild")
+    if rhbuild.startswith("6"):
+        rhbuild = "quincy"
+    elif rhbuild.startswith("5"):
+        rhbuild = "pacific"
+    else:
+        raise Exception("Unsupported version")
+
+    # Setup ssh key to all nodes
+    setup_ssh_keys(installer, ceph_nodes)
+
+    # Get registry credentials
+    registry, reg_username, reg_password = get_registry_details(config)
+
+    # Setup a private registry on installer node
+    setup_private_container_registry(
+        installer=installer,
+        nodes=ceph_nodes,
+        registry=registry,
+        reg_username=reg_username,
+        reg_password=reg_password,
+        build_type=rhbuild,
+        private_reg_username=PRIVATE_REG_USER,
+        private_reg_password=PRIVATE_REG_PASS,
+        docker_reg_image=docker_reg_image,
+        images=dashboard_images,
+    )
+
+    # Get private registry catalog
+    out = get_private_registry_image(
+        installer, username=PRIVATE_REG_USER, password=PRIVATE_REG_PASS
+    )
+    repos = out.get("repositories")
+
+    # Pull images to nodes
+    for ceph in ceph_nodes:
+        Registry(ceph).login(
+            registry=f"{installer.hostname}:{PORT}",
+            username=PRIVATE_REG_USER,
+            password=PRIVATE_REG_PASS,
+        )
+        for img in repos:
+            Container(ceph).pull(image=f"{installer.hostname}:{PORT}/{img}")
+
+    for config_key, image in registry_image.items():
+        img_reg = image.replace(f"{registry}/", "")
```

Comment:
why this replacement is required? is this specific to registry?

Reply:
We need to get `rhceph/rhceph-5-dashboard-rhel8` to pass it to image name: 

 ceph-custom-image-verify-h0hwsi-node1-installer:5000/rhceph/rhceph-5-dashboard-rhel8

While setting the config for mgr:
ceph config set mgr mgr/cephadm/container_image_grafana ceph-custom-image-verify-h0hwsi-node1-installer:5000/rhceph/rhceph-5-dashboard-rhel8:latest

This replacement is to get the rest of image name expect for the registry name from image. 

Let me know if it's still not clear.

Reply:
This is incorrect way to do it. It will fail in case any other registry is provided. Use `regex` to get registry, repo and tag

Reply:
Sure, using regex got the repo and tag.